lib/ssl/src/tls_gen_connection.erl

```diff
@@ -77,36 +77,68 @@
+        {Starter, Error} ->
+            Error
     end;
-
 start_fsm(Role, Host, Port, Socket,
-          {#{erl_dist := true, sender_spawn_opts := SenderOpts}, _, Trackers} = Opts,
+          {#{erl_dist := true = ErlDist, sender_spawn_opts := SenderOpts}, _, Trackers} = Opts,
 	  User, {CbModule, _, _, _, _} = CbInfo, 
-	  Timeout) -> 
-    try
+	  Timeout) ->
         SenderOpts1 = [{priority, max} | proplists:delete(priority, SenderOpts)],
-        {ok, DynSup} = tls_connection_sup:start_child_dist([]),
-        {ok, Sender} = tls_dyn_connection_sup:start_child(DynSup, sender, [[{spawn_opt, SenderOpts1}]]),
-	{ok, Pid} = tls_dyn_connection_sup:start_child(DynSup, receiver, [Role, Sender, Host, Port, Socket,
-                                                                 Opts, User, CbInfo]),
-	{ok, SslSocket} = ssl_gen_statem:socket_control(?MODULE, Socket, [Pid, Sender], CbModule, Trackers),
-        ssl_gen_statem:handshake(SslSocket, Timeout)
-    catch
-	error:{badmatch, {error, _} = Error} ->
-	    Error
+    Starter = start_connection_tree(User, ErlDist, [[{spawn_opt, SenderOpts1}]],
+                                    Role, [Host, Port, Socket,Opts, User, CbInfo]),
+    receive
+        {Starter, SockReceiver, SockSender} ->
+            socket_control(Socket, SockReceiver, SockSender, CbModule, Trackers, Timeout);
+        {Starter, Error} ->
+            Error
+    end.
+
+start_connection_tree(User, ErlDist, SenderOpts, Role, ReceiverOpts) ->
+    DynConnectionSupChild = case ErlDist of
+                           true ->
+                               start_child_dist;
+                           false ->
+                               start_child
+                       end,
+    StartConnectionTree =
+        fun() ->
+                case tls_connection_sup:DynConnectionSupChild([]) of
+                    {ok, DynSup} ->
+                        case tls_dyn_connection_sup:start_child(DynSup, sender, SenderOpts) of
+                            {ok, Sender} ->
+                                case tls_dyn_connection_sup:start_child(DynSup, receiver,
+                                                                        [Role, Sender | ReceiverOpts]) of
+                                    {ok, Receiver} ->
+                                        User ! {self(), Receiver, Sender};
+                                    {error, Error} ->
+                                        User ! {self(), Error},
+                                        exit(DynSup, shutdown)
+                                end;
+                            {error, Error} ->
+                                User ! {self(), Error},
+                                exit(DynSup, shutdown)
+                        end;
+                    {error, Error} ->
+                        User ! {self(), Error}
+                end
+        end,
+    spawn(StartConnectionTree).
```

Comment:
I suggest something like below in order to avoid a dynamic call  with the function name in a variable
```suggestion
dyn_connection_sup_child(true,ArgList) ->
    tls_connection_sup:start_child_dist(ArgList);
dyn_connection_sup_child(false,ArgList) ->
    tls_connection_sup:start_child(ArgList);
    
start_connection_tree(User, ErlDist, SenderOpts, Role, ReceiverOpts) ->
    StartConnectionTree =
        fun() ->
                case dyn_connection_sup_child(ErlDist, []) of
                    {ok, DynSup} ->
                        case tls_dyn_connection_sup:start_child(DynSup, sender, SenderOpts) of
                            {ok, Sender} ->
                                case tls_dyn_connection_sup:start_child(DynSup, receiver,
                                                                        [Role, Sender | ReceiverOpts]) of
                                    {ok, Receiver} ->
                                        User ! {self(), Receiver, Sender};
                                    {error, Error} ->
                                        User ! {self(), Error},
                                        exit(DynSup, shutdown)
                                end;
                            {error, Error} ->
                                User ! {self(), Error},
                                exit(DynSup, shutdown)
                        end;
                    {error, Error} ->
                        User ! {self(), Error}
                end
        end,
    spawn(StartConnectionTree).
```

Reply:
Dynamic function calls can be really useful, but in this case, it is not really that valuable so we can make it along the suggested lines. I thought I made the choice really obvious though!